ddtrace/_trace/pin.py

```diff
@@ -0,0 +1,237 @@
+from typing import TYPE_CHECKING  # noqa:F401
+from typing import Any  # noqa:F401
+from typing import Dict  # noqa:F401
+from typing import Optional  # noqa:F401
+
+import wrapt
+
+import ddtrace
+from ddtrace.internal.utils.deprecations import DDTraceDeprecationWarning
+from ddtrace.vendor.debtcollector import deprecate
+
+from ..internal.logger import get_logger
+
+
+log = get_logger(__name__)
+
+
+# To set attributes on wrapt proxy objects use this prefix:
+# http://wrapt.readthedocs.io/en/latest/wrappers.html
+_DD_PIN_NAME = "_datadog_pin"
+_DD_PIN_PROXY_NAME = "_self_" + _DD_PIN_NAME
+
+
+class Pin(object):
+    """Pin (a.k.a Patch INfo) is a small class which is used to
+    set tracing metadata on a particular traced connection.
+    This is useful if you wanted to, say, trace two different
+    database clusters.
+
+        >>> conn = sqlite.connect('/tmp/user.db')
+        >>> # Override a pin for a specific connection
+        >>> pin = Pin.override(conn, service='user-db')
+        >>> conn = sqlite.connect('/tmp/image.db')
+    """
+
+    __slots__ = ["tags", "tracer", "_target", "_config", "_initialized"]
+
+    def __init__(
```

Comment:
Can we revisit the design of the Pin class and its usage. Currently the Pin object is used in some integrations but is missing or half supported in others. It would be nice if we were consistent. We shouldn't do this in this PR but it's something to think about.